crates/assists/src/handlers/generate_enum_match_method.rs

```diff
@@ -34,79 +35,284 @@ use crate::{
+        },
+    )
+}
+
+// Assist: generate_enum_as_method
+//
+// Generate an `as_` method for an enum variant.
+//
+// ```
+// enum Value {
+//  Number(i32),
+//  Text(String)$0,
+// }
+// ```
+// ->
+// ```
+// enum Value {
+//  Number(i32),
+//  Text(String),
+// }
+//
+// impl Value {
+//     fn as_text(&self) -> Option<&String> {
+//         if let Self::Text(v) = self {
+//             Some(v)
+//         } else {
+//             None
+//         }
+//     }
+// }
+// ```
+pub(crate) fn generate_enum_as_method(acc: &mut Assists, ctx: &AssistContext) -> Option<()> {
+    let variant = ctx.find_node_at_offset::<ast::Variant>()?;
+    let variant_name = variant.name()?;
+    let parent_enum = ast::Adt::Enum(variant.parent_enum());
+    let variant_kind = variant_kind(&variant);
+
+    let fn_name = format!("as_{}", &to_lower_snake_case(variant_name.text()));
+
+    // Return early if we've found an existing new fn
+    let impl_def = find_struct_impl(&ctx, &parent_enum, &fn_name)?;
+
+    let field_type = variant_kind.single_field_type()?;
+    let (pattern_suffix, bound_name) = variant_kind.binding_pattern()?;
+
+    let target = variant.syntax().text_range();
+    acc.add(
+        AssistId("generate_enum_as_method", AssistKind::Generate),
+        "Generate an `as_` method for an enum variant",
+        target,
+        |builder| {
+            let vis = parent_enum.visibility().map_or(String::new(), |v| format!("{} ", v));
+            let method = format!(
+                "    {}fn {}(&self) -> Option<&{}> {{
+        if let Self::{}{} = self {{
+            Some({})
+        }} else {{
+            None
+        }}
+    }}",
+                vis,
+                fn_name,
+                field_type.syntax(),
+                variant_name,
+                pattern_suffix,
+                bound_name,
+            );
 
-            builder.insert(start_offset, buf);
+            add_method_to_adt(builder, &parent_enum, impl_def, &method);
         },
     )
 }
 
+fn add_method_to_adt(
+    builder: &mut AssistBuilder,
+    adt: &ast::Adt,
+    impl_def: Option<ast::Impl>,
+    method: &str,
+) {
+    let mut buf = String::with_capacity(method.len() + 2);
+    if impl_def.is_some() {
+        buf.push('\n');
+    }
+    buf.push_str(method);
+
+    let start_offset = impl_def
+        .and_then(|impl_def| find_impl_block_end(impl_def, &mut buf))
+        .unwrap_or_else(|| {
+            buf = generate_impl_text(&adt, &buf);
+            adt.syntax().text_range().end()
+        });
+
+    builder.insert(start_offset, buf);
+}
+
+enum VariantKind {
+    Unit,
+    /// Tuple with a single field
+    NewtypeTuple {
+        ty: Option<ast::Type>,
+    },
+    /// Tuple with 0 or more than 2 fields
+    Tuple,
+    /// Record with a single field
+    NewtypeRecord {
+        field_name: Option<ast::Name>,
+        field_type: Option<ast::Type>,
+    },
+    /// Record with 0 or more than 2 fields
+    Record,
+}
```

Comment:
This seems like the kind of functionality for which I'd expect an abstraction to already exist within Rust-Analyzer. cc/ @Veykril

If it turns out this is something new, at least for now it'd still be good to have this live under `assists/utils.rs` so other assists can also make use of this.

Reply:
There is no abstraction like this yet to my knowledge, most assists only required `ast::StructKind` which doesn't discern between single field record/tuples.

Reply:
Also from the looks of it, it doesn't really make sense to keep the `Option` in the variants here. Every accessor just propogates the `None` values which are then propogated at every call site as well. I think making `variant_kind` the constructor return an `Option<VariantKind>` instead when any of those are `None` makes more sense.